src/Common/src/TypeSystem/Common/TypeDesc.cs

```diff
@@ -465,5 +481,25 @@ public virtual MethodDesc GetFinalizer()
         {
             return null;
         }
+
+        /// <summary>
+        /// Gets a value indicating whether this type has generic variance (the definition of the type
+        /// has a generic parameter that is co- or contravariant).
+        /// </summary>
+        public bool HasVariance
+        {
+            get
+            {
+                return (GetTypeFlags(TypeFlags.HasGenericVariance | TypeFlags.HasGenericVarianceComputed) & TypeFlags.HasGenericVariance) != 0;
+            }
+        }
+
+        /// <summary>
+        /// Gets the kind of this type.
+        /// </summary>
+        public abstract TypeKind Variety
```

Comment:
> The switch over  Category  becomes ugly

Having the method implemented as just a switch would look better to me. It routes the execution quickly to the place it needs to go - instead of the current mix of ifs and switch:

```
switch (thisType.Category)
{
case TypeFlags.GenericParameter:
     return thisType.CanCastToGenericParameter(otherType);

case TypeFlags.Array:
case TypeFlags.SzArray:
     if (otherType.IsDefType)
         return thisType.CanCastToClassOrInterface(otherType);
     return thisType.CanCastToParametrizedType(otherType);

case TypeFlags.ByRef:
case TypeFlags.Pointer:
     return thisType.CanCastToParametrizedType(otherType);

case TypeFlags.Interface:
    return thisType.CanCastToInterface(otherType);

default:
    Debug.Assert(otherType.IsDefType());
    return thisType.CanCastToClass(otherType);
}
```

> TypeFlags don't distinguish between SzArray and Array

This is fine to fix. I think it is just a historic artifact as the type system developed.